linkerd/app/gateway/src/lib.rs

```diff
@@ -20,49 +33,144 @@ pub struct Config {
+    metrics: &metrics::Proxy,
+    span_sink: Option<mpsc::Sender<oc::Span>>,
+    drain: drain::Watch,
 ) -> impl svc::NewService<
-    inbound::HttpGatewayTarget,
-    Service = impl tower::Service<
-        http::Request<http::BoxBody>,
-        Response = http::Response<http::BoxBody>,
-        Error = impl Into<Error>,
-        Future = impl Send,
-    > + Send
+    GatewayConnection,
+    Service = impl tower::Service<I, Response = (), Error = impl Into<Error>, Future = impl Send>
+                  + Send
                   + 'static,
 > + Clone
        + Send
 where
-    P: profiles::GetProfile<NameAddr> + Clone + Send + 'static,
+    I: io::AsyncRead + io::AsyncWrite + io::PeerAddr + Send + Sync + Unpin + 'static,
+    P: profiles::GetProfile<NameAddr> + Clone + Send + Sync + Unpin + 'static,
     P::Future: Send + 'static,
     P::Error: Send,
-    O: svc::NewService<outbound::http::Logical, Service = S> + Clone + Send + 'static,
-    S: tower::Service<http::Request<http::BoxBody>, Response = http::Response<http::BoxBody>>
+    O: svc::NewService<outbound::http::Logical, Service = OSvc>,
+    O: Clone + Send + Sync + Unpin + 'static,
+    OSvc: tower::Service<http::Request<http::BoxBody>, Response = http::Response<http::BoxBody>>
         + Send
         + 'static,
-    S::Error: Into<Error>,
-    S::Future: Send + 'static,
+    OSvc::Error: Into<Error>,
+    OSvc::Future: Send + 'static,
 {
-    svc::stack(NewGateway::new(outbound, local_id))
-            .check_new_service::<gateway::Target, http::Request<http::BoxBody>>()
-            .push(profiles::discover::layer(
-                profiles,
-                Allow(allow_discovery),
-            ))
-            .instrument(
-                |g: &inbound::HttpGatewayTarget| debug_span!("gateway", target = %g.target, v = %g.version),
+    let http_server = {
+        // Cache an HTTP gateway service for each destination and HTTP version.
+        //
+        // The destination is determined from the transport header or, if none was
+        // present, the request URI.
+        //
+        // The client's ID is set as a request extension, as required by the
+        // gateway. This permits gateway services (and profile resolutions) to be
+        // cached per target, shared across clients.
+        let gateway = svc::stack(NewGateway::new(outbound, local_id))
+            .push(profiles::discover::layer(profiles, Allow(allow_discovery)))
+            .instrument(|h: &HttpTarget| debug_span!("gateway", target = %h.target, v = %h.version))
+            .push_on_response(
+                svc::layers()
+                    .push(svc::layer::mk(svc::SpawnReady::new))
+                    .push(
+                        metrics
+                            .stack
+                            .layer(metrics::StackLabels::inbound("http", "gateway")),
+                    )
+                    .push(svc::FailFast::layer("Gateway", inbound.dispatch_timeout))
```

Comment:
nit/tioli: does it make sense to have a stack helper for "`SpawnReady`. stack metrics, and `FailFast`"? seems like they almost always occur together

Reply:
we need to create a spawn ready layer that does the right thing or wait for tower 0.5